### DISCUSSION TITLE
A2DP half duplex - possible?

### DISCUSSION THREAD

**Discussion by vorosj** [pschatzmann/arduino-audio-tools#1001] [Q&A] [answered]
Is it possible to use Bluetooth A2DP profile half duplex (like a Push To Talk WalkieTalkie)?
I know there are profiles like headset, but the corresponding codec is just brutal.

I am about to use an ESP32-A1S as a simple SDR transceiver frontend. In FT8 mode, it first receives for several seconds, and I like to transfer the received sound stream to the PC, to a program like WSJT-X. Then the PC (the WSJT-X program) sends sound stream to the ESP32.
Is it possible? Or it disconnects at every direction swap?
One more question: is it possible to run a Bluetooth serial port connection parallel on the same device?

If the above is not possible (or just too hard to write such a program), maybe better to use Wifi RAW audio? For example UDP packets. In this case, how to handle the stream on the PC side? It has to be accessible to existing programs like a virtual sound card. How to do that? I searched for solutions, unfortunately found only hints for Linux, but I prefer Windows. It surprised me, because even a simple project, like a Wifi microphone for Skype etc. needs it. 

Bluetooth audio has the advantage that it appears as sound source or sink for audio programs, no interface program required. However, not sure, how the codec degrades the signal (the FT8 decoder is working even below noise level, overlapping strong and weak signals etc, maybe the codec kills important details of the signal).

Wifi is capable to transfer without codecs, but I don't know how to make or find a stream<->[virtual soundcard] program.

Please, point me to the right direction.

**Comment by pschatzmann** [comment, marked as answer]
No, it is not: https://github.com/pschatzmann/ESP32-A2DP/discussions/283 and sending FT8 over an audio codec will just not work!

I think you really need to decide if you transmit FT8 or audio: I would first try to stick with FT8 and do the decoding and encoding on the desktop. Did you already find out if USB/Serial is too slow for this ?

If you want to write your own solution, I suggest that you have a look at the provided communication examples.You try to can compile and run the sketches written with the audio tools on the desktop as well. 

If you just want to deal with audio (record, convert, stream) on the desktop, I suggest that you have a look at ffmpeg

Unfortunately I am not using Windows, so you will need to find this out yourself, but usually the concepts are pretty similar.

**Reply by vorosj** [reply]
Thank you for the fast reply.

_sending FT8 over an audio codec will just not work_
The FT8_CN app on Android has Bluetooth option, and on this video it seems to work: https://www.youtube.com/watch?v=ksZIzT51usU
I have no Icom705, and the video won't tell the connection profile. 
The FT8 signal (considering the SSB demodulation was realised on the uC side) is low bandwith, only 3kHz, considering the Nyquist limit even 8ks/sec is enough with 16bit resolution. Maybe it is possible without a codec.

_Did you already find out if USB/Serial is too slow for this ?_

A composite usb connection using audio and serial is perfect, I already wrote such a program using an STM32 uC. (not Arduino project)
But I am interested in a wireless solution as well. The ESP32-A1S with the built-in 24bit codec seems ideal for a simple QRP radio, no usb, but wireless choices.

Wifi seems easier on the uC side. I am familiar with embedded programming, but not on the Windows PC side. Unfortunately the FT8 programs like WSJT-X have no direct stream options, can't avoid the Windows sound system. Typical use is connecting the radio using a USB soundcard., or - in case of a PC based SDR - pipeing with a sw like Virtual Audio Cable. 

So - considering a Wifi stream (either RAW or more complicated) - I need something similar to VAC, but with the Wifi stream on one side.
I googled a lot for "wifi headset", "ESP32 microphone" and many other keywords, but to my surprise no results. There are microphone codes for ESP32, but no Windows interface to became a choosable input device. Really strange, I suspected it's a common need for a lot of projects using audio. Maybe I missed something obvious?

**Comment by pschatzmann** [comment]
Like I have written in my A2DP Project: A2DP on the ESP32 expects PCM data that will be encoded via SBC and you can't really replace this.

Wifi is not really relevant here but the communication protocols which are IP or UDP. That's why I was recommending ffmpeg which provides a lot of functionality...

If you look for a Bluetooth solution, you will need to implement some more additional protocols:  Headset (HSP) or Hands-Free (HFP) profiles which are supported by the ESP32, but you will need to dig into the related APIs yourself

**Reply by pschatzmann** [reply]
ps. with such low rates: you can even look into Bluetooth Serial.

**Comment by vorosj** [comment]
With any data transfer solution (udp or whatever) the last step is unclear for me: how to make it accessible to programs as a windows audio source? For example, I write a Wifi microphone with inmp441 and esp32, and it sends the data with udp, so far so good. But how to use the stream, for example as a microphone is Skype, or any other programs? How to make it selectable in the list of available audio sources?  I attach a picture, if it's unclear what I mean. Is there a program, that do it for me? I am very surpised, that it's not mentioned very often in forums, as it seems to be an obvious feature when dealing with audio devices. Or I missed something trivial, misunderstood something?
![audio_source](https://github.com/pschatzmann/arduino-audio-tools/assets/12972823/50497b7c-167e-4482-853e-ac808edfc929)

**Comment by pschatzmann** [comment]
Just google and look at the command line parameters of ffmpeg. 
If you can use Bluetooth Serial, it would be just the related serial port

**Reply by vorosj** [reply]
I searched a lot, but didn't find a solution with ffmpeg alone for a virtual microphone. Under Linux, pactl (part of PulseAudio) creates a virtual microphone, and that can be a target for ffmpeg. The whole thing is really complicated even under Linux. 
I found a possible solution: VBAN (VoiceMeeter), they defined a packet format, that I can easily send or receive, and hopefully their program will do the job. At least I understand, what they are talking about.
With ffmpeg, pulseadio etc there are overwhelming details, and I don't understand the main concept.

**Comment by pschatzmann** [comment]
I thought you talk windows: I would expect that you can just play audio from udp by defining a udp source and don't need to care too much how the output is working. Did you try just to play an audio file to check if the output is working a expected...

**Reply by vorosj** [reply]
Thank you very much for help so much.

Probably we dont really understand each other. I think that's because I don't know how Windows handles devices, streams etc. under hood. Just playing an udp stream doesn't help me.

My device has to talk to existing programs, that is the problem.

Those programs only input is a microphone (a list of available Windows devices, as seen on the picture above), and the only output is a list of available Windows sound devices. 
I clearly understand, how udp works, and I can send or receive udp packets from an ESP32 to _my_ program on a PC (that I write in Java or C#), but don't know, how to connect to an existing program like Skype, WSJT-X etc. It needs to emulate somehow a windows sound device that appears on that lists. 
 I am familiar with embedded programming, but absolute beginner in programming under Windows operating system.

For me, Arduino is a concept to hide complicated details, make it easy for the user. Your library is a good example, hides ugly details of i2s, dma, registers in the hw etc. Just pick some classes, fill some parameters, start an enjoy.
And then, when I try to use it for a simple job, making a for example a wireless wifi microphone and use it like a usual microphone - no simple solution. Linux style commands, billion parameters, interactions with other command line objects etc.

**Reply by pschatzmann** [reply]
Yes, but usually there is a default input and output. Playing audio via the command line is usually just sending it to the default output, if you don't specify anything else..

**Reply by vorosj** [reply]
Wow, that was fast. :-)
How can this default output appears as a choosable microphon for an existing program? That's unclear for me. What I found is really complicated: Python scripts using ffmepeg, pulseaudio etc programs.

**Reply by pschatzmann** [reply]
I hope you mean speaker: sound is played on a speaker and not a microphone. I would expect that whatever your machine is that it will automatically played via the built in speakers. If you have multiple output devices, on linux or os/x you can change this in the os if it should go somewhere else. The same is valid for the audio input side.

I guess in Windows you can define and change the default devices as well..

**Reply by vorosj** [reply]
(I am not only asking dumb questions instead od working - I just write an udp test audio stream for the VBAN pipeline)

**Reply by pschatzmann** [reply]
Well, I guess you skipped a topic: you can only play audio to an audio destination. FT8 is not audio and I would expect that you need to decode this somewhere first

**Reply by vorosj** [reply]
_I hope you mean speaker: sound is played on a speaker and not a microphone._ 

Sure, but it's a different case... :-)

Consider I make a wireless microphone using ESP32 (I2S->udp),. On the PC side an existing program, for example Skype, has a choosable mic input.
So I need an interface program, a "virtual microphone" program. This "virtual microphone" _receives_ the udp stream, and _sends_ it somehow to Skype, appears for Skype as a choosable microphone, passes the stream to Skype..

**Reply by pschatzmann** [reply]
I think you can pretty much forget about this scenario: You don't want to write your own windows device drivers!

**Comment by vorosj** [comment]
_Well, I guess you skipped a topic: you can only play audio to an audio destination. FT8 is not audio and I would expect that you need to decode this somewhere first_

Ok, I am lamer, but not _that_ much... :-)
In this case, the signal is 0...3kHz bandwith audio, that is the output of my receiver. This audio signal is the input for the FT8 decoder, in this case the WSJT-X program, that is running on the PC. The WSJT-X doesn't have direct stream interface (shame). My job is feeding the WSJT-X program with the audio signal, it has to appear as a mic input for the WSJT-X program (like the picture above).
And vica-versa, the WSJT-X provides an audio stream (a choosable speaker output) that I have to send to the ESP32.

I already did it using USB. I defined a composite USB device, two way audio plus a serial port. It's working very well. (the serial port has an independent job, kind of remote control, sets the band, frequency, mode of the radio etc)

Now I am about to do the same, but using wireless sonnection. The serial port is not a problem, but the audio i/o.

**Reply by pschatzmann** [reply]
Hmm, I wanted to propose USB, but did not because the regular ESP32 would not offer that.
I guess the only other way that might be supported by the OS is to use a2dp with hfp hsp, as indicated in the link of the original answer.

Unfortunaltey, there are no easy to use libraries for this, only some ESP32 example code

**Reply by vorosj** [reply]
My usb version is based on a STM32F4 chip and TinyUSB. This is possible with ESP32S3 as well, but now I try to do it Wireless.
Hope I can make the VBAN work tomorrow, if that's succesful, I search for a similar open source version.
Thank you for helping.

**Reply by pschatzmann** [reply]
In theory, yes - the last time I checked there were still some (timing) issues that prevented audio from working properly on the ESP32S3.
ps. Just double checked: the issue has been resolved...
